### ISSUE TITLE
[Bug] Does not finish loading (sidebar panel flickers and does not show restored tabs)

### ISSUE THREAD

**Issue by mYnDstrEAm** [piroor/treestyletab#3329]
## Abstract

This has been happening for a long time. However, now it looks like it doesn't ever finish loading while earlier it did so eventually when letting it load for long (without switching the panel in between which seems to restart the loading). Maybe it's because the count of tabs grew even larger and because there's now two windows instead of just one.

I was trying to restore the session in Firefox with History->Restore previous session.

I noticed that it could load the tabs when the theme is changed to "No Decoration" as described in the (related?) issue #3130. It did finish loading when that was set. However, the entire tree structure was lost which is also described in other issues. That's why I restored the session with the "Tab Session Manager" (where I increased the time per tab restore to something over 1 second). The same problem occurred in the middle of when it was restoring. My workaround then was to
* use History->Restore previous session to restore the partially restored session
* export the TabSessionManager session of the session that I want to restore (in the addon settings)
* edit this file to remove all tabs that were already restored and import this session
* restore this session to a new window

This seemed to have worked, except that I then tried to move all the tabs from the new window over into the window where they originally were located in, which made the same problem appear again. However, later I noticed it was far fewer tabs than before anyway, maybe because the tabs are sorted differently in the exported session file.

----

It's a big problem if the addon doesn't just load very long but can (and frequently does) lose tree structure or tabs, I think it's something that could be prevented with some small code changes. For example, maybe it too quickly decides to reload or can get broken by some problematic tab for example instead of excluding it to show it in an "could not restore xyz" message or has an unstable way to load the theme.

Help appreciated, this is a big problem, what else could I try (workarounds welcome too)? Now it partly restored the session with the merged windows and the restoring looks different from other times: it does not flicker and, instead of showing all restored tabs at once once it finishes loading if it ever does, it slowly adds one tree after another to the panel of one of the two windows (the second only showed a grey area for long but loaded the tabs after waiting for long...however many tree structures were lost because I had to use the workaround described above and later I noticed that the window had far fewer tabs so I had to restore the window again with TSM). One thing I did **differently is to not click on "Tree Style Tab" in the sidebar and switch it to "Bookmarks" and back** which probably restarts the loading. To fix this issue described here, please check what is being done when the shown sidebar is switched (how switching back and forth affects the restoring).

**Edit**: so I tried to reopen the 2nd window session with TSM and it added one tab after another but after a while the area turned just grey with no tabs visible anymore. When trying to reload the panel by switching to bookmarks and back to TST the same problem (flickering) occurs. Also when browsing few of the open tabs, Javascript seems to be disabled. Some script may have frozen. It shouldn't do much loading because when restoring the tabs, the pages aren't opened (they are unloaded and only load once one clicks on them). After a while, always after roughly at the same number of tabs but not exactly (so it's probably not due to a specific tab breaking it), the sidebar panel turns grey instead of displaying the tabs and when I click on close to show the number of open tabs it displays a number lower than the tabs in the previous session and the number doesn't increase anymore (so it stopped restoring tabs).

**Edit 2**: I got it restore all tabs by decreasing the "Delay when tree restoration" in the Tab Session Manager to a very small number and trying it a few more times. I disconnected from the Internet, maybe that helped. However, it also didn't load the favicons. If there is a way to make it load all favicons afterwards from a local file, let me know.

* Another issue is that TabSessionManager should be recommended by the addon in a way that the user reads that because otherwise people can lose all their tabs, which happened many tabs looking through the issues, maybe I should open a new issue for that. It should be a high priority that people to make absolutely sure people don't lose their sessions.
* A further one is enabling the user to find out why it loads this long or prevent that this occurs (see for example #3012). It makes no sense that this process is that resource-intensive, hence it's likely a problem with code (of the addon or other software).

## Steps to reproduce

 1. Start Firefox with clean profile.
 2. Install TST.
 3. Open many tabs in two windows and quit firefox
 4. Restore with History->Restore previous session

## Expected result
Restore all the tabs with their tree structure in under 5 minutes.

## Actual result
Flickering (flashing) sidebar and tabs are not restored in a way that they show in the panel even when letting it load for hours (waiting for long used to be the solution).

https://user-images.githubusercontent.com/10249049/235348166-c2271fe5-1f17-478b-8348-3214a4227195.mp4

## Environment

 * Platform (OS): Debian 11
* Version of Firefox: 102.10.0esr (latest in Debian11 repos)
* Version (or revision) of Tree Style Tab: 3.9.15 (latest, no update available)
* Hardware is good enough [but without graphics card](https://unix.stackexchange.com/questions/583635/where-can-i-find-a-comparison-of-graphics-cards-using-free-software-drivers-are)

**Comment by irvinm** [comment]
@mYnDstrEAm you haven't said anything about your configuration in terms of CPU and memory which could have a lot of implications on performance.  However, I have used TSM and TST to restore large sessions in the past ... and as long as you are using a large enough delay for TST support and restore them as discarded tabs, you should have no problems doing 500+ tabs.

In terms of:

> If there is a way to make it load all favicons afterwards from a local file, let me know.

I had made an addon a year ago that does this online.  It loads your tabs (one at a time) just to get the favicon, then discards it again.  Useful only for this specific scenario of restoring a session from TSM.  (https://addons.mozilla.org/en-US/firefox/addon/reload-discarded-tabs/) [Mozilla ask to allow creating a discarded tab with a favicon:  [Bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1475240)]

**Comment by mYnDstrEAm** [comment]
It worked in the past too until now. I just had to wait for long (always this same flickering occurred) but that didn't help this time. Now I could set the delay to something like 50 ms and it restored the session fine. I don't know if disconnecting from the Internet was a reason for why it worked this time and/or whether the favicons are part of the reason why it crashed while restoring.

Aren't the favicons stored locally? I think they are because they can be compressed in the settings and because it would have to make requests to all the sites while restoring just for the favicons which is something you'd like to avoid for reasons like privacy and load.

The favicons are downloaded the first time you visit a website so Firefox should keep them stored until you don't have the tab in your current or prior session anymore. Maybe TSM could basically ensure the favicons of its sessions (or of all its named sessions or the session marked as current session or depending on some setting) are kept if it doesn't already do so (for example protect all favicons of domains opened during the last two years and a TSM feature that restores favicons during restoration for any tabs whose domains you haven't opened during that time).

I wondered about how favicons are stored and requested and generally the requests during restoration due to privacy issues before but don't think that I have already asked about that.

The addon seems like a useful workaround, but not in my use-case. It should only load the favicon once per domain, not reload every tab if it currently does so.

It seems like this is more of an issue with TSM. I made an issue there too later which is linked above.
* What would happen if I close the session and then restore the session with _History->Restore previous session_ in Firefox without disconnecting from the Internet?
* Would that make requests to all the domains or all the websites or properly use some local file?
* If it uses a local file and it got emptied could I restore the favicons from a backup?
* When I load one domain for which there are many tabs, is there a way to make it update the favicon in the displayed tabs for all of the tabs at once at that point?

These may become separate issues:

* Has there already been a request about for example showing all tabs of one specific domain in TST without altering the tree structure? For example, showing all github tabs so you can go through them at once, but then afterwards (or even in parallel) still have the tree view of tabs with the previous structure in TST?
* Why is the unloading functionality of "Tab Unloader for Tree Style Tab" not part of TST? It's very useful and it's not good to have such useful in separate addons, for example because most users will never learn that these exist or that functionality could exist or be useful.

**Comment by irvinm** [comment]
> Aren't the favicons stored locally? I think they are because they can be compressed in the settings and because it would have to make requests to all the sites while restoring just for the favicons which is something you'd like to avoid for reasons like privacy and load.

Yes, Firefox stores the favicons in favicons.sqlite (https://support.mozilla.org/en-US/kb/profiles-where-firefox-stores-user-data).  However, the way that tabs.create() with the "discarded" property, the favicon is not set\loaded until the first time the tab is made active.  Therefore, whether the favicon is stored locally or needs to be retrieved doesn't matter until the tab is loaded.

> The addon seems like a useful workaround, but not in my use-case. It should only load the favicon once per domain, not reload every tab if it currently does so.

So, while it would be easy enough to just load\discard 1 tab per domain, this would not work as it won't refresh all of the tabs for that domain once the first one is loaded.  (You should be able to test that by restoring a session and activating 1 tab for a domain and seeing the rest are not updated at that point)

> It seems like this is more of an issue with TSM. I made an issue there too later which is linked above.

I don't think there is much TSM can do with this Firefox limitation.  If the tab is loaded as discarded, there is no way to set the favicon.  If you load the tab during the restore, the favicon issue would be resolved but you would likely run Firefox out of memory depending on the size of the session.

> What would happen if I close the session and then restore the session with History->Restore previous session in Firefox without disconnecting from the Internet?

At the end of the day, this basically takes the "previous" sessionstore.jsonlz4 from the backup folder and makes it the current one.  This is actually one method to back up your sessions that does not even require TSM.  I run a script that backs up the most recent file in the "sessionstore-backups" folder every 15 minutes.  To restore, just close Firefox, place the backed up file into the root folder, rename it to sessionstore.jsonlz4, and start Firefox.  It will restore that session with all of it favicons intact.

You can try this out for yourself ... it works really well and you don't have to worry about how quickly TSM works or the favicon issue.

**Comment by mYnDstrEAm** [comment]
> However, the way that tabs.create() with the "discarded" property, the favicon is not set\loaded until the first time the tab is made active. Therefore, whether the favicon is stored locally or needs to be retrieved doesn't matter until the tab is loaded.

With loaded do you mean web requests to reopen the tab or for example loading from cache or only loading local data?

It would be a problem if session restoration makes requests to all the tabs, not only due to privacy, performance problems and possible crashes but also because websites could think that it's a DOS or something.

If you really meant that and there is no way around that is there an issue about that fore Firefox? Seems like something they should fix asap.

>So, while it would be easy enough to just load\discard 1 tab per domain, this would not work as it won't refresh all of the tabs for that domain once the first one is loaded.

Yes, I know that's currently not the case but maybe it's possible that the addon makes that happen during restoration. If that's not possible (and maybe even if it is), it should get fixed at Firefox not only because of privacy but also because it would speed up performance by a lot. Is there already an issue for this?

> If the tab is loaded as discarded, there is no way to set the favicon.

If that is what the bug above is about: is there a way to make them finally solve this? For example by asking about it on reddit about clarifying why this would be very useful in the bug comments. I don't see how this would difficult to do, you'd only need to load the favicon from the local file so maybe it's intentionally not being done (why isn't it already done right from when the sqlite file was first implemented)? I thought I already subscribed to the bug but haven't and did now.

>but you would likely run Firefox out of memory depending on the size of the session.

This (favicon loading causing memory issues) may have been the cause for my and many others' crashes and session losses. Something should be done about that I guess such as new options (like "restore without favicon" or "only load favicons from local file") and/or a warning when the addon crashed or the user opens a large session/window. However, while it used a lot of RAM it seemed like I still had enough (note that TST, TSM and FF should also work with only 2GB of RAM).

>This is actually one method to back up your sessions that does not even require TSM. I run a script that backs up the most recent file in the "sessionstore-backups" folder every 15 minutes.

Yes but it should be only a fallback because it can fail and hence you need to backups but most people don't do proper backups and even most who do probably don't know about that folder.

>To restore, just close Firefox, place the backed up file into the root folder, rename it to sessionstore.jsonlz4, and start Firefox. It will restore that session with all of it favicons intact.
>You can try this out for yourself ... it works really well and you don't have to worry about how quickly TSM works or the favicon issue.

So is this the workaround concerning the crashes and the favicon issue: restore while disconnected from the Internet to prevent it from retrieving favicons, then close the browser and restore via History->Restore previous session to make it also load the favicons? If that doesn't load the favicons: could you do that but after closing the browser copy over a backed up favicons.sqlite?

**Comment by irvinm** [comment]
> If that is what the bug above is about: is there a way to make them finally solve this? For example by asking about it on reddit about clarifying why this would be very useful in the bug comments. I don't see how this would difficult to do, you'd only need to load the favicon from the local file so maybe it's intentionally not being done (why isn't it already done right from when the sqlite file was first implemented)? I thought I already subscribed to the bug but haven't and did now.

I have already made this comment ... 5 years ago.  https://bugzilla.mozilla.org/show_bug.cgi?id=1475240#c2

> This (favicon loading causing memory issues) may have been the cause for my and many others' crashes and session losses. Something should be done about that I guess such as new options (like "restore without favicon" or "only load favicons from local file") and/or a warning when the addon crashed or the user opens a large session/window. However, while it used a lot of RAM it seemed like I still had enough (note that TST, TSM and FF should also work with only 2GB of RAM).

I think I have to disagree with you on this one.  I doubt favicon loading is the issue and I am pretty sure most browsers are not intended to work with only 2GB of RAM.  (If you do, it should still work if you have enough swapfile space, but the performance will be very poor.)

> Yes but it should be only a fallback because it can fail and hence you need to backups but most people don't do proper backups and even most who do probably don't know about that folder.

I don't disagree with this.  This solution isn't for the casual user.  I know there are other items they are tracking to have a better means of "session management" at the addon level, but they are several years old and probably not going to happen.  In an ideal world, the addon API would allow storing sessions (including favicons) and allow the user to selectively restore sessions and\or just tabs however they see fit.  (Like the old "session manager" addon before FF57)

**Comment by mYnDstrEAm** [comment]
I think this can be closed but I'm not sure if it's a TSM issue or (also) a TST issue.
New issues for the things mentioned here: #3332 #3333 and #3334

----

>I doubt favicon loading is the issue

Ok but then I don't know what may have caused it and why it worked when I disconnected from the Internet and all the favicons weren't loaded.

>In an ideal world, the addon API would allow storing sessions (including favicons)

Let me know if / whenever there is an issue at Firefox about this. When storing multiple sessions, no favicons should be stored twice. This could be used by TSM and if such is already possible it would be a TSM issue to enable offline from-local-file favicon loading. Maybe I'll create an issue for that or ask at the TSM repo. Also I don't yet understand why it loaded the favicons earlier when I also used to load the tabs as unloaded but didn't load them offline / the time restoration worked, why is that?

**Comment by piroor** [comment]
@mYnDstrEAm Could you upload video of the debugger console instead of the sidebar? https://github.com/piroor/treestyletab/wiki/How-to-inspect-tree-of-tabs#how-to-start-debugger-for-tree-style-tab Something fatal error message may appear in the console when TST reloads repeatedly. Video looks a better way to collect logs on your case because logs are cleared when TST reloads itself.

**Comment by mYnDstrEAm** [comment]
Will do so the next time this happens.

The main problem here was that it couldn't restore the session and lost tab trees. Restoring while offline with a very small "Delay when tree restoration" in TSM is a workaround albeit the favicons aren't loaded. This flickering problem occurred for a long time and repeatedly so I suspect other users have the same problem and that this has something to do with why it used to take so long to restore.

Edit:
* Concerning the favicons, one workaround is to use [TST Colored Tabs](https://addons.mozilla.org/en-US/firefox/addon/tst-colored-tabs/) to set colors for domains as an alternative to favicons.

* Also left a comment at [the bugzilla issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1475240) asking the favicon issue to get solved.

**Comment by mYnDstrEAm** [comment]
Okay I still have this problem and it's very annoying. I still need to use this workaround (disconnecting machine from Internet, small TSM delay, TSM restore) when restoring a session and it looks like I can't use _History->Restore previous session_ anymore.

I think this is because I now have two windows, earlier I had the same problem but I got it to work when I only have one window and let it load for long. Here's some output from the addon's about:debugging console:

`managed storage is not provided  Configs.js:207:27`

I think the number of the previous error repeated many times, probably as many as there are tabs if I wouldn't have quit the browser before it reached that number. Only opening the browser there's this:

```
unreachable code after return statement
tree.js:815:8
managed storage is not provided Configs.js:207:27
TSTAPI.initAsBackend: failed to send "ready" message to "multipletab@piro.sakura.ne.jp": Error: Could not establish connection. Receiving end does not exist.
id moz-extension://{id}/common/tst-api.js:568
tst-api.js:583:15
managed storage is not provided 2 Configs.js:207:27
```

Restoring:

```
TSTAPI.initAsBackend: failed to send "ready" message to "multipletab@piro.sakura.ne.jp": Error: Could not establish connection. Receiving end does not exist.
    id moz-extension://{id}/common/tst-api.js:568
tst-api.js:583:15
managed storage is not provided 4 Configs.js:207:27
Uncaught (in promise) undefined 2 scroll.js:336:5
Uncaught (in promise) undefined 105 scroll.js:336:5
Uncaught (in promise) undefined scroll.js:336:5
Uncaught (in promise) undefined scroll.js:336:5
Uncaught (in promise) undefined scroll.js:336:5
Uncaught (in promise) undefined scroll.js:336:5
Uncaught (in promise) undefined scroll.js:336:5
Error: Error: tracked tabs are not same to pulled tabs, for the window 237. Rebuilding...
    syncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:189
    timer moz-extension://{id}/sidebar/sidebar-tabs.js:137
    setTimeout handler*reserveToSyncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:135
    syncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:195
    timer moz-extension://{id}/sidebar/sidebar-tabs.js:137
    setTimeout handler*reserveToSyncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:135
    syncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:195
    timer moz-extension://{id}/sidebar/sidebar-tabs.js:137
    setTimeout handler*reserveToSyncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:135
    syncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:195
    timer moz-extension://{id}/sidebar/sidebar-tabs.js:137
    setTimeout handler*reserveToSyncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:135
    syncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:195
    timer moz-extension://{id}/sidebar/sidebar-tabs.js:137
    setTimeout handler*reserveToSyncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:135
    syncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:195
    timer moz-extension://{id}/sidebar/sidebar-tabs.js:137
    setTimeout handler*reserveToSyncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:135
    syncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:195
    timer moz-extension://{id}/sidebar/sidebar-tabs.js:137
    setTimeout handler*reserveToSyncTabsOrder moz-extension://{id}/sidebar/sidebar-tabs.js:135
sidebar-tabs.js:189:15
Uncaught (in promise) undefined 5 scroll.js:336:5
Uncaught (in promise) TypeError: tab.$TST.element is null
    calculateScrollDeltaForTab moz-extension://{id}/sidebar/scroll.js:111
    isTabInViewport moz-extension://{id}/sidebar/scroll.js:138
    notifyOutOfViewTab moz-extension://{id}/sidebar/scroll.js:403
    onBackgroundMessage moz-extension://{id}/sidebar/scroll.js:575
    promise callback*onBackgroundMessage moz-extension://{id}/sidebar/scroll.js:564
    dispatchWithDetails moz-extension://{id}/extlib/EventListenerManager.js:64
    dispatch moz-extension://{id}/extlib/EventListenerManager.js:43
    onConnectionMessage moz-extension://{id}/sidebar/background-connection.js:127
    onConnectionMessage moz-extension://{id}/sidebar/background-connection.js:113
3 scroll.js:111:25
Uncaught (in promise) undefined 46 scroll.js:336:5
Uncaught Error: Attempt to postMessage on disconnected port
    onFrame moz-extension://{id}/common/sidebar-connection.js:134
    setTimeout handler*sendMessageToPort moz-extension://{id}/common/sidebar-connection.js:146
    sendMessage moz-extension://{id}/common/sidebar-connection.js:99
    onRemoved moz-extension://{id}/background/api-tabs-listener.js:833
sidebar-connection.js:134:12
Uncaught (in promise) undefined 5 scroll.js:336:5
{above error many times}
```

After the above the session seemed fully restored but for the first time I had the problem that I got this normal Firefox error when trying to open a tab (some tabs.js or similarly named script failed and I stopped it but I think that was at an earlier try but it could also have been this try and have something to do with that):

```
Hmm. We’re having trouble finding that site.
We can’t connect to the server at {domain}.
If that address is correct, here are three other things you can try: {...}
```

I could still open websites in another instance of Firefox (different profile).

After successfully restoring via offline TSM restoration where I can see and use all tabs (without favicons) in both windows there is this in the console:

```
unreachable code after return statement
tree.js:815:8
managed storage is not provided Configs.js:207:27
Uncaught Error: Attempt to postMessage on disconnected port
    onFrame moz-extension://{id}/common/sidebar-connection.js:134
2 sidebar-connection.js:134:12
TSTAPI.initAsBackend: failed to send "ready" message to "multipletab@piro.sakura.ne.jp": Error: Could not establish connection. Receiving end does not exist.
    id moz-extension://{id}/common/tst-api.js:568
tst-api.js:583:15
TSTAPI.initAsBackend: failed to send "ready" message to "tst-colored-tabs@murz": Error: Could not establish connection. Receiving end does not exist.
    id moz-extension://{id}/common/tst-api.js:568
tst-api.js:583:15
managed storage is not provided 4 Configs.js:207:27
Uncaught (in promise) Error: Invalid tab ID: 7145 3 undefined
Uncaught (in promise) Error: Invalid tab ID: 7151 3 undefined
Uncaught (in promise) Error: Invalid tab ID: 7171 3 undefined
```

**Comment by mYnDstrEAm** [comment]
Just some further output in case that helps eventual debugging of the problem (it happened again):

When restoring with History->Restore previous session and then switching the panel this output blinks in the console (I think it's one output then reload then the same output again):
`managed storage is not provided Config.js:207:27`
With "switching the panel" I mean changing the panel from Tree Style Tab to Bookmarks and back.

I then tried it again without switching the panel and I think it was while offline (at least partly):

```
unreachable code after return statement
tree.js:815:8
Uncaught (in promise) Error: Invalid tab ID: x61 2 undefined
Uncaught (in promise) Error: Invalid tab ID: x61 undefined
Uncaught (in promise) Error: Invalid tab ID: x62 undefined
Uncaught (in promise) undefined Tab.js:1371:16
Uncaught (in promise) undefined successor-tab.js:188:19
Uncaught (in promise) Error: Invalid tab ID: x62 2 undefined
Uncaught (in promise) Error: Invalid tab ID: x62 undefined
```

(x is a number) In an earlier try there was a popup about Tab.js has stopped responding Continue or Stop script (I think I tried both). I have two windows that's why some outputs appeared twice (with a 2). After the above output I switched the panel again and the output stayed there but `managed storage not provided` is added beneath it. Then it loads (left bar _without_ flickering). When I switch the panel in the second window `managed storage not provided (2)` is added again. It even loaded the favicons in one of the windows despite that I was offline. Then these two outputs appear:
```
Uncaught (in promise) Error: Invalid tab ID: x18 undefined
Uncaught (in promise) Error: Invalid tab ID: x19 undefined
```
and then
```
managed storage is not provided Configs.js:207:27
Uncaught (in promise) Error: Invalid tab ID: x64 3 undefined
```
It looked like it would have restored but when I come back online I can't open the site as Firefox (latest version for Debian) just displays `Hmm. We’re having trouble finding that site.` So I tried it again once or a few times with Restore previous session and it restored the session. However, the tab trees weren't restored in the window that has more tabs. Maybe that's because the previous session is the session I tried to restore previously and that already didn't load the trees. So now the trees of the second window are flat.
